File: Shields/Shields/Shields.m
//
//  Shields.m
//  Shields
//
//  Created by Mac on 2021/5/21.
//

#import "Shields.h"
#import "ShieldsMessage.h"
#import "ShieldsModel.h"
#import <WebKit/WKWebView.h>

@interface Shields()
@property (nonatomic,strong)ShieldsLabel *label;
@property (nonatomic,strong)ShieldsMessage *message;
@end

@implementation Shields
- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        /// 配置UI
        [self setupUI:ShieldsDefault];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame style:(ShieldsStyle)style {
    self = [super initWithFrame:frame];
    if (self) {
        /// 配置UI
        [self setupUI:style];
    }
    return self;
}

#pragma mark - 配置UI
- (void)setupUI:(ShieldsStyle)style {
    /// label
    CGFloat labelW = CGRectGetWidth(self.frame) * 0.4;
    ShieldsLabel *label = [[ShieldsLabel alloc] initWithFrame:CGRectMake(0, 0, labelW, CGRectGetHeight(self.frame)) style:style];
    [self addSubview:label];
    self.label = label;
    
    CGFloat messageW = CGRectGetWidth(self.frame) - labelW;
    /// message
    ShieldsMessage *message = [[ShieldsMessage alloc] initWithFrame:CGRectMake(labelW, 0, messageW, CGRectGetHeight(self.frame))];
    [self addSubview:message];
    self.message = message;
}

- (void)configLabel:(NSString *)label message:(NSString *)message {
    [self configLabel:label message:message labelColor:nil messageColor:nil];
}

- (void)configLabel:(NSString *)label message:(NSString *)message labelColor:(UIColor *)labelColor messageColor:(UIColor *)messageColor {
    [self configLabel:label message:message labelBackgroundColor:nil messageBackgroundColor:nil labelColor:labelColor messageColor:messageColor logo:nil logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message labelColor:(UIColor *)labelColor {
    [self configLabel:label message:message labelBackgroundColor:nil messageBackgroundColor:nil labelColor:labelColor messageColor:nil logo:nil logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message messageColor:(UIColor *)messageColor {
    [self configLabel:label message:message labelBackgroundColor:nil messageBackgroundColor:nil labelColor:nil messageColor:messageColor logo:nil logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message labelBackgroundColor:(UIColor *)labelBackgroundColor messageBackgroundColor:(UIColor *)messageBackgroundColor {
    [self configLabel:label message:message labelBackgroundColor:labelBackgroundColor messageBackgroundColor:messageBackgroundColor labelColor:nil messageColor:nil logo:nil logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message labelBackgroundColor:(UIColor *)labelBackgroundColor {
    [self configLabel:label message:message labelBackgroundColor:labelBackgroundColor messageBackgroundColor:nil labelColor:nil messageColor:nil logo:nil logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message messageBackgroundColor:(UIColor *)messageBackgroundColor {
    [self configLabel:label message:message labelBackgroundColor:nil messageBackgroundColor:messageBackgroundColor labelColor:nil messageColor:nil logo:nil logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message labelBackgroundColor:(UIColor *)labelBackgroundColor messageBackgroundColor:(UIColor *)messageBackgroundColor labelColor:(UIColor *)labelColor messageColor:(UIColor *)messageColor {
    [self configLabel:label message:message labelBackgroundColor:labelBackgroundColor messageBackgroundColor:messageBackgroundColor labelColor:labelColor messageColor:messageColor logo:nil logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message labelBackgroundColor:(UIColor *)labelBackgroundColor messageBackgroundColor:(UIColor *)messageBackgroundColor labelColor:(UIColor *)labelColor messageColor:(UIColor *)messageColor logo:(UIImage *)logo {
    [self configLabel:label message:message labelBackgroundColor:labelBackgroundColor messageBackgroundColor:messageBackgroundColor labelColor:labelColor messageColor:messageColor logo:logo logoWidth:0.0 logoPosition:CGPointZero];
}

- (void)configLabel:(NSString *)label message:(NSString *)message labelBackgroundColor:(UIColor *)labelBackgroundColor messageBackgroundColor:(UIColor *)messageBackgroundColor labelColor:(UIColor *)labelColor messageColor:(UIColor *)messageColor logo:(UIImage *)logo logoWidth:(CGFloat)logoWidth logoPosition:(CGPoint)logoPosition {
    if (self.label == nil && self.message == nil) {
        return ;
    }
    
    if (labelBackgroundColor != nil) {
        /// label背景颜色
        self.label.backgroundColor = labelBackgroundColor;
    }
    
    if (messageBackgroundColor != nil) {
        /// message背景颜色
        self.message.backgroundColor = messageBackgroundColor;
    }
    
    if (labelColor != nil) {
        /// 标签颜色
        [self.label configLabel:label color:labelColor];
    }
    
    if (messageColor != nil) {
        /// 信息颜色
        [self.message configMessage:message color:messageColor];
    }
    
    if (logo != nil) {
        [self.label configLogo:logo logoWidth:logoWidth logoPosition:logoPosition];
    }
}

#pragma mark - Json文件读取配置
- (void)configWithJsonFilePath:(NSString *)jsonPath {
    if (jsonPath.length != 0) {
        [self configJsonWithData:[NSData dataWithContentsOfFile:jsonPath]];
    }
}

- (void)configWithJsonFileURL:(NSURL *)jsonURL {
    if (jsonURL != nil) {
        [self configJsonWithData:[NSData dataWithContentsOfURL:jsonURL]];
    }
}

- (void)configJsonWithData:(NSData *)data {
    NSError *error = nil;
    NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:&error];
    if (error != nil) {
        NSLog(@"%@",error.localizedDescription);
        return ;
    }
    
    [self configWithModel:[ShieldsModel modelWithDict:dict]];
}

- (void)configWithModel:(ShieldsModel *)model {
    [self configLabel:model.label message:model.message labelBackgroundColor:model.labelBackgroundColor messageBackgroundColor:model.messageBackgroundColor labelColor:model.labelColor messageColor:model.messageColor logo:[UIImage imageNamed:model.logoName] logoWidth:model.logoWidth logoPosition:model.logoPosition];
}

#pragma mark - 配置属性
- (void)setMessageWidth:(CGFloat)messageWidth {
    _messageWidth = messageWidth;
    if (self.message != nil) {
        CGPoint center = self.center;
        self.message.frame = CGRectMake(self.message.frame.origin.x, self.message.frame.origin.y, messageWidth, CGRectGetHeight(self.message.frame));
        [self.message layoutSubviews];
        
        self.frame = CGRectMake(self.frame.origin.x, self.frame.origin.y, CGRectGetWidth(self.label.frame)+messageWidth, CGRectGetHeight(self.frame));
        self.center = center;
    }
}

- (void)setCornerRadius:(CGFloat)cornerRadius {
    _cornerRadius = cornerRadius;
    self.layer.cornerRadius = cornerRadius;
    self.layer.masksToBounds = YES;
}

@end

File: Swift/Shields/Shields/Shields/Shields.swift
//
//  Shields.swift
//  Shields
//
//  Created by 纽扣软件 on 2021/5/24.
//

import UIKit

class Shields: UIView {
    //
    private var label: ShieldsLabel = ShieldsLabel()
    private var message: ShieldsMessage = ShieldsMessage()
    
    //
    var messageWidth: CGFloat {
        get {
            15.0
        }
        set {
            let center: CGPoint = self.center
            message.frame = CGRect(x: self.message.frame.origin.x, y: self.message.frame.origin.y, width: newValue, height: self.message.frame.height)
            message.layoutSubviews()
            
            self.frame = CGRect(x: self.frame.origin.x, y: self.frame.origin.y, width: self.label.frame.width+newValue, height: self.frame.height)
            self.center = center
        }
    }
    
    var cornerRadius: CGFloat {
        get {
            0.0
        }
        
        set {
            layer.cornerRadius = newValue
            layer.masksToBounds = true
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        /// 配置UI
        setupUI(style: .ShieldsDefault)
    }
    
    init(frame: CGRect, style: ShieldsLabel.ShieldsStyle) {
        super.init(frame: frame)
        
        /// 配置UI
        setupUI(style: style)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

//MARK:- 配置UI
extension Shields {
    private func setupUI(style: ShieldsLabel.ShieldsStyle) {
        /// label
        let labelW: CGFloat = self.frame.width * 0.4;
        let label: ShieldsLabel = ShieldsLabel(frame: CGRect(x: 0, y: 0, width: labelW, height: self.frame.height), style: style)
        addSubview(label)
        self.label = label;
        
        let messageW: CGFloat = self.frame.width - labelW;
        /// message
        let message: ShieldsMessage = ShieldsMessage(frame: CGRect(x: labelW, y: 0, width: messageW, height: self.frame.height))
        addSubview(message)
        self.message = message;
    }
}

//MARK:- 基本配置
extension Shields {
    func configLabel(label: String?, message: String?) {
        configLabel(label: label, message: message, labelColor: nil, messageColor: nil)
    }
    
    func configLabel(label: String?, message: String?, labelColor: UIColor?, messageColor: UIColor?) {
        configLabel(label: label, message: message, labelBackgroundColor: nil, messageBackgroundColor: nil, labelColor: labelColor, messageColor: messageColor)
    }
    
    func configLabel(label: String?, message: String?, labelColor: UIColor?) {
        configLabel(label: label, message: message, labelColor: labelColor, messageColor: nil)
    }
    
    func configLabel(label: String?, message: String?, messageColor: UIColor?) {
        configLabel(label: label, message: message, labelColor: nil, messageColor: messageColor)
    }
    
    func configLabel(label: String?, message: String?, labelBackgroundColor: UIColor?) {
        configLabel(label: label, message: message, labelBackgroundColor: labelBackgroundColor, messageBackgroundColor: nil, labelColor: nil, messageColor: nil)
    }
    
    func configLabel(label: String?, message: String?, messageBackgroundColor: UIColor?) {
        configLabel(label: label, message: message, labelBackgroundColor: nil, messageBackgroundColor: messageBackgroundColor, labelColor: nil, messageColor: nil)
    }
    
    func configLabel(label: String?, message: String?, labelBackgroundColor: UIColor?, messageBackgroundColor: UIColor?) {
        configLabel(label: label, message: message, labelBackgroundColor: labelBackgroundColor, messageBackgroundColor: messageBackgroundColor, labelColor: nil, messageColor: nil)
    }
    
    func configLabel(label: String?, message: String?, labelBackgroundColor: UIColor?, messageBackgroundColor: UIColor?, labelColor: UIColor?, messageColor: UIColor?) {
        configLabel(label: label, message: message, labelBackgroundColor: labelBackgroundColor, messageBackgroundColor: messageBackgroundColor, labelColor: labelColor, messageColor: messageColor, logo: nil)
    }
    
    func configLabel(label: String?, message: String?, labelBackgroundColor: UIColor?, messageBackgroundColor: UIColor?, labelColor: UIColor?, messageColor: UIColor?, logo: UIImage?) {
        configLabel(label: label, message: message, labelBackgroundColor: labelBackgroundColor, messageBackgroundColor: messageBackgroundColor, labelColor: labelColor, messageColor: messageColor, logo: logo, logoWidth: 0.0, logoPosition: CGPoint())
    }
    
    func configLabel(label: String?, message: String?, labelBackgroundColor: UIColor?, messageBackgroundColor: UIColor?, labelColor: UIColor?, messageColor: UIColor?, logo: UIImage?, logoWidth: CGFloat?, logoPosition: CGPoint?) {
        if let labelBackgroundColor = labelBackgroundColor {
            self.label.backgroundColor = labelBackgroundColor
        }
        
        if let messageBackgroundColor = messageBackgroundColor {
            self.message.backgroundColor = messageBackgroundColor
        }

        
        
        if let label = label {
            self.label.configLabel(str: label)
            if let labelColor = labelColor {
                self.label.configLabel(str: label, color: labelColor)
            }
        }

        if let message = message  {
            self.message.configMessage(str: message)
            if let messageColor = messageColor {
                self.message.configMessage(str: message, color: messageColor)
            }
        }
        
        
        if let logo = logo {
            self.label.configLogo(img: logo)
            if let logoWidth = logoWidth {
                self.label.configLogo(img: logo, logoWidth: logoWidth)
                if let logoPosition = logoPosition {
                    self.label.configLogo(img: logo, logoWidth: logoWidth, logoPosition: logoPosition)
                }
            }
        }
    }
}

//MARK:- Json文件读取配置
extension Shields {
    func configWithJsonFilePath(jsonPath: String) {
        if jsonPath.count != 0 {
            do {
                let data = try Data.init(contentsOf: URL(fileURLWithPath: jsonPath))
                configJsonWithData(data: data)
            } catch {
                print(error.localizedDescription)
            }
        }
    }
    
    func configWithJsonFileURL(jsonURL: URL) {
        if jsonURL.absoluteString.count != 0 {
            do {
                let data = try Data.init(contentsOf: jsonURL)
                configJsonWithData(data: data)
            } catch {
                print(error.localizedDescription)
            }
        }
    }
    
    private func configJsonWithData(data: Data) {
        do {
            let dict = try JSONSerialization.jsonObject(with: data, options: .mutableContainers)
            configWithModel(model: ShieldsModel().modelWithDict(dict: dict as! Dictionary<String, Any>))
        } catch {
            print(error.localizedDescription)
            return
        }
    }
    
    private func configWithModel(model: ShieldsModel) {
        configLabel(label: model.label, message: model.message, labelBackgroundColor: model.labelBackgroundColor, messageBackgroundColor: model.messageBackgroundColor, labelColor: model.labelColor, messageColor: model.messageColor, logo: UIImage(named: model.logoName), logoWidth: model.logoWidth, logoPosition: model.logoPosition)
    }
}
